cloudcix_primitives/firewallns.py

```diff
@@ -93,14 +93,98 @@ def snat_rule(nat):
     return rule_line
 
 
+def validate_rules(rules, set_names, messages, msg_index):
+    messages_list = []
+    errors = []
+    valid_rules = True
+    validated = True
+    for rule in rules:
+        controller = FirewallNamespace(rule)
+        success, errs = controller()
+        if success is False:
+            valid_rules = False
+            errors.extend(errs)
+    if valid_rules is False:
+        validated = False
+        messages_list.append(f'{messages[msg_index]} {";".join(errors)}')
+
+    # check if the rule has any sets in source, destination or ports,
+    # if so then check that set is defined in sets
+    errors = []
+    valid_set_elements = True
+    for rule in rules:
+        rule_sets = []
+        # collect the sets from the rules that starts with `@`
+        if type(rule['source']) is list:
+            rule_sets.extend([item.strip('@') for item in rule['source'] if '@' in item])
+        if type(rule['destination']) is list:
+            rule_sets.extend([item.strip('@') for item in rule['destination'] if '@' in item])
+        if type(rule['port']) is list:
+            rule_sets.extend([item.strip('@') for item in rule['port'] if '@' in item])
+        # now check if each rule_set is supplied in sets(set_names)
+        for rule_set in rule_sets:
+            if rule_set not in set_names:
+                valid_set_elements = False
+                errors.append(f'{rule_set} not found in the supplied sets')
+    if valid_set_elements is False:
+        validated = False
+        messages_list.append(f'{messages[msg_index + 1]} {";".join(errors)}')
+
+    return validated, messages_list
+
+
+def validate_nats(nats, messages, msg_index):
+    messages_list = []
+    errors = []
+    valid_nats = True
+    validated = True
+    nats_list = nats['dnats'] + nats['snats']
+    for nat in nats_list:
+        controller = FirewallNAT(nat)
+        success, errs = controller()
+        if success is False:
+            valid_nats = False
+            errors.extend(errs)
+    if valid_nats is False:
+        validated = False
+        messages_list.append(f'{messages[msg_index]} {";".join(errors)}')
+    return validated, messages_list
+
+
+def validate_sets(sets, messages, msg_index):
+    messages_list = []
+    set_names = []
+    valid_sets = True
```

Comment:
Why are two booleans required? 

From what I am reading both will either be True or both will be False. I am not seeing a scenario where one is True and the other is False